From 70304fe7155a47132a4dc2a755a5a24358efb59a Mon Sep 17 00:00:00 2001
From: Jacob Dahl <37091262+dakejahl@users.noreply.github.com>
Date: Mon, 20 May 2024 12:35:29 -0600
Subject: [PATCH] [mavlink] Parameter to always start on USB (#22234)

* usb: Added parameter to enable always starting mavlink on USB.

    Refactored cdcacm_init into a module and added a paramter to allow always starting mavlink on
    USB, also added a paramter to choose the mode. The current default behavior is to wait and listen
    for data on USB and auto-detect the protocol (mavlink, nsh, ublox). This results in the mavlink
    stream not starting until something else on the mavlink network sends a packet first. The new
    default behavior is to always start mavlink.

    Added parameters
    MAV_USB_ENABLE -- default 1 (always start mavlink on USB)
    MAV_USE_MODE -- default 3 (onboard)

* added 3 retries for opening serial port in mavlink, removed sleep before sercon

* added DRIVERS_CDCACM_AUTOSTART to ark-v6x default.px4board

* added CONFIG_DRIVERS_CDCACM_AUTOSTART=y to default.px4board for boards with CONFIG_CDCACM in their nsh/defconfig

* format

* remove PGA460 from COMMON_DISTANCE_SENSOR to save flash

* remove LIS2MDL from COMMON_MAGNETOMETER to save flash

* disable CONFIG_DRIVERS_CDCACM_AUTOSTART for fmu-v5 protected.px4board

* moved and renamed parameters, removed mode logic in mavlink

* changed parameter names, added mode none

* remove parameters from mavlink
---
 ROMFS/px4fmu_common/init.d/rcS                |   8 +
 boards/airmind/mindpx-v2/default.px4board     |   1 +
 boards/ark/fmu-v6x/default.px4board           |   2 +
 boards/atl/mantis-edu/default.px4board        |   1 +
 boards/bitcraze/crazyflie/default.px4board    |   1 +
 boards/bitcraze/crazyflie21/default.px4board  |   1 +
 boards/cuav/nora/default.px4board             |   1 +
 boards/cuav/x7pro/default.px4board            |   1 +
 boards/cubepilot/cubeorange/default.px4board  |   1 +
 .../cubepilot/cubeorangeplus/default.px4board |   1 +
 boards/cubepilot/cubeyellow/default.px4board  |   1 +
 .../diatone/mamba-f405-mk2/default.px4board   |   1 +
 boards/flywoo/gn-f405/default.px4board        |   1 +
 boards/hkust/nxt-dual/default.px4board        |   1 +
 boards/hkust/nxt-v1/default.px4board          |   1 +
 boards/holybro/durandal-v1/default.px4board   |   1 +
 boards/holybro/kakutef7/default.px4board      |   1 +
 boards/holybro/kakuteh7/default.px4board      |   1 +
 boards/holybro/kakuteh7mini/default.px4board  |   1 +
 boards/holybro/kakuteh7v2/default.px4board    |   1 +
 boards/holybro/pix32v5/default.px4board       |   1 +
 boards/matek/h743-mini/default.px4board       |   1 +
 boards/matek/h743-slim/default.px4board       |   1 +
 boards/matek/h743/default.px4board            |   1 +
 boards/modalai/fc-v1/default.px4board         |   1 +
 boards/modalai/fc-v2/default.px4board         |   1 +
 boards/mro/ctrl-zero-classic/default.px4board |   1 +
 boards/mro/ctrl-zero-f7-oem/default.px4board  |   1 +
 boards/mro/ctrl-zero-f7/default.px4board      |   1 +
 boards/mro/ctrl-zero-h7-oem/default.px4board  |   1 +
 boards/mro/ctrl-zero-h7/default.px4board      |   1 +
 boards/mro/pixracerpro/default.px4board       |   1 +
 boards/mro/x21-777/default.px4board           |   1 +
 boards/mro/x21/default.px4board               |   1 +
 boards/nxp/fmuk66-e/default.px4board          |   1 +
 boards/nxp/fmuk66-v3/default.px4board         |   1 +
 boards/omnibus/f4sd/default.px4board          |   1 +
 boards/px4/fmu-v2/default.px4board            |   1 +
 boards/px4/fmu-v3/default.px4board            |   1 +
 boards/px4/fmu-v4/default.px4board            |   1 +
 boards/px4/fmu-v4pro/default.px4board         |   1 +
 boards/px4/fmu-v5/default.px4board            |   1 +
 boards/px4/fmu-v5/protected.px4board          |   1 +
 boards/px4/fmu-v5x/default.px4board           |   1 +
 boards/px4/fmu-v6c/default.px4board           |   1 +
 boards/px4/fmu-v6u/default.px4board           |   1 +
 boards/px4/fmu-v6x/default.px4board           |   1 +
 boards/px4/fmu-v6xrt/default.px4board         |   1 +
 boards/raspberrypi/pico/default.px4board      |   1 +
 boards/siyi/n7/default.px4board               |   1 +
 .../smartap-airlink/default.px4board          |   1 +
 boards/spracing/h7extreme/default.px4board    |   1 +
 boards/thepeach/k1/default.px4board           |   1 +
 boards/thepeach/r1/default.px4board           |   1 +
 boards/uvify/core/default.px4board            |   1 +
 platforms/nuttx/src/px4/common/CMakeLists.txt |   1 -
 platforms/nuttx/src/px4/common/px4_init.cpp   |   6 -
 .../nuttx/src/px4/common/px4_layer.cmake      |   1 -
 .../src/px4/common/px4_protected_layers.cmake |   1 -
 .../src/px4/common/px4_userspace_init.cpp     |   6 -
 src/drivers/cdcacm_autostart/CMakeLists.txt   |  40 ++
 src/drivers/cdcacm_autostart/Kconfig          |   6 +
 .../cdcacm_autostart/cdcacm_autostart.cpp     | 663 ++++++++++++++++++
 .../cdcacm_autostart/cdcacm_autostart.h       | 124 ++++
 .../cdcacm_autostart_params.c                 |  68 ++
 src/modules/mavlink/mavlink_main.cpp          |  26 +-
 src/systemcmds/serial_passthru/Kconfig        |   2 +-
 67 files changed, 982 insertions(+), 25 deletions(-)
 create mode 100644 src/drivers/cdcacm_autostart/CMakeLists.txt
 create mode 100644 src/drivers/cdcacm_autostart/Kconfig
 create mode 100644 src/drivers/cdcacm_autostart/cdcacm_autostart.cpp
 create mode 100644 src/drivers/cdcacm_autostart/cdcacm_autostart.h
 create mode 100644 src/drivers/cdcacm_autostart/cdcacm_autostart_params.c

diff --git a/ROMFS/px4fmu_common/init.d/rcS b/ROMFS/px4fmu_common/init.d/rcS
index b9d46ce27d..c0b895a8fe 100644
--- a/ROMFS/px4fmu_common/init.d/rcS
+++ b/ROMFS/px4fmu_common/init.d/rcS
@@ -439,6 +439,14 @@ else
 	# Must be started after the serial config is read
 	rc_input start $RC_INPUT_ARGS
 
+	# Manages USB interface
+	if ! cdcacm_autostart start
+	then
+		sercon
+		echo "Starting MAVLink on /dev/ttyACM0"
+		mavlink start -d /dev/ttyACM0
+	fi
+
 	#
 	# Play the startup tune (if not disabled or there is an error)
 	#
diff --git a/boards/airmind/mindpx-v2/default.px4board b/boards/airmind/mindpx-v2/default.px4board
index cb56388c36..5f4bcd2b0c 100644
--- a/boards/airmind/mindpx-v2/default.px4board
+++ b/boards/airmind/mindpx-v2/default.px4board
@@ -12,6 +12,7 @@ CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_ICM20948=y
 CONFIG_DRIVERS_IMU_INVENSENSE_MPU6000=y
diff --git a/boards/ark/fmu-v6x/default.px4board b/boards/ark/fmu-v6x/default.px4board
index 4018e493a2..1209adb9dc 100644
--- a/boards/ark/fmu-v6x/default.px4board
+++ b/boards/ark/fmu-v6x/default.px4board
@@ -8,12 +8,14 @@ CONFIG_BOARD_SERIAL_TEL2="/dev/ttyS4"
 CONFIG_BOARD_SERIAL_TEL3="/dev/ttyS1"
 CONFIG_BOARD_SERIAL_TEL4="/dev/ttyS3"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_BAROMETER_BMP388=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_HEATER=y
diff --git a/boards/atl/mantis-edu/default.px4board b/boards/atl/mantis-edu/default.px4board
index 58fcec76d7..fd76e4d6ca 100644
--- a/boards/atl/mantis-edu/default.px4board
+++ b/boards/atl/mantis-edu/default.px4board
@@ -2,6 +2,7 @@ CONFIG_BOARD_TOOLCHAIN="arm-none-eabi"
 CONFIG_BOARD_ARCHITECTURE="cortex-m7"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_MAIERTEK_MPC2520=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_ICM20602=y
 CONFIG_DRIVERS_MAGNETOMETER_ISENTEK_IST8310=y
diff --git a/boards/bitcraze/crazyflie/default.px4board b/boards/bitcraze/crazyflie/default.px4board
index e7e9705f2d..02b70b58c3 100644
--- a/boards/bitcraze/crazyflie/default.px4board
+++ b/boards/bitcraze/crazyflie/default.px4board
@@ -3,6 +3,7 @@ CONFIG_BOARD_ARCHITECTURE="cortex-m4"
 CONFIG_BOARD_CONSTRAINED_FLASH=y
 CONFIG_BOARD_CONSTRAINED_MEMORY=y
 CONFIG_BOARD_COMPILE_DEFINITIONS="-Wno-narrowing"
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_BAROMETER_LPS25H=y
 CONFIG_DRIVERS_DISTANCE_SENSOR_VL53L0X=y
 CONFIG_DRIVERS_GPS=y
diff --git a/boards/bitcraze/crazyflie21/default.px4board b/boards/bitcraze/crazyflie21/default.px4board
index aa55b50621..0c268b6983 100644
--- a/boards/bitcraze/crazyflie21/default.px4board
+++ b/boards/bitcraze/crazyflie21/default.px4board
@@ -4,6 +4,7 @@ CONFIG_BOARD_CONSTRAINED_FLASH=y
 CONFIG_BOARD_CONSTRAINED_MEMORY=y
 CONFIG_BOARD_COMPILE_DEFINITIONS="-Wno-narrowing"
 CONFIG_DRIVERS_BAROMETER_BMP388=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DISTANCE_SENSOR_VL53L1X=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_BOSCH_BMI088_I2C=y
diff --git a/boards/cuav/nora/default.px4board b/boards/cuav/nora/default.px4board
index f312504ee5..2a0e40e37e 100644
--- a/boards/cuav/nora/default.px4board
+++ b/boards/cuav/nora/default.px4board
@@ -7,6 +7,7 @@ CONFIG_BOARD_SERIAL_TEL2="/dev/ttyS3"
 CONFIG_DRIVERS_ADC_ADS1115=y
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_COMMON_BAROMETERS=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
diff --git a/boards/cuav/x7pro/default.px4board b/boards/cuav/x7pro/default.px4board
index 99e831c129..b2212d5f15 100644
--- a/boards/cuav/x7pro/default.px4board
+++ b/boards/cuav/x7pro/default.px4board
@@ -6,6 +6,7 @@ CONFIG_BOARD_SERIAL_TEL1="/dev/ttyS1"
 CONFIG_BOARD_SERIAL_TEL2="/dev/ttyS3"
 CONFIG_DRIVERS_ADC_ADS1115=y
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
diff --git a/boards/cubepilot/cubeorange/default.px4board b/boards/cubepilot/cubeorange/default.px4board
index 91ddcda596..e3c84b1838 100644
--- a/boards/cubepilot/cubeorange/default.px4board
+++ b/boards/cubepilot/cubeorange/default.px4board
@@ -11,6 +11,7 @@ CONFIG_DRIVERS_BAROMETER_MS5611=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/cubepilot/cubeorangeplus/default.px4board b/boards/cubepilot/cubeorangeplus/default.px4board
index 837876ee7d..d755eea8d8 100644
--- a/boards/cubepilot/cubeorangeplus/default.px4board
+++ b/boards/cubepilot/cubeorangeplus/default.px4board
@@ -11,6 +11,7 @@ CONFIG_DRIVERS_BAROMETER_MS5611=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/cubepilot/cubeyellow/default.px4board b/boards/cubepilot/cubeyellow/default.px4board
index 45cf93a38d..62cf898af9 100644
--- a/boards/cubepilot/cubeyellow/default.px4board
+++ b/boards/cubepilot/cubeyellow/default.px4board
@@ -12,6 +12,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/diatone/mamba-f405-mk2/default.px4board b/boards/diatone/mamba-f405-mk2/default.px4board
index 32ec4fd47b..e8ee5ec656 100644
--- a/boards/diatone/mamba-f405-mk2/default.px4board
+++ b/boards/diatone/mamba-f405-mk2/default.px4board
@@ -8,6 +8,7 @@ CONFIG_BOARD_SERIAL_URT6="/dev/ttyS2"
 CONFIG_BOARD_SERIAL_TEL2="/dev/ttyS1"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_BMP280=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_ICM20602=y
diff --git a/boards/flywoo/gn-f405/default.px4board b/boards/flywoo/gn-f405/default.px4board
index b016b0db9a..54d17325bc 100644
--- a/boards/flywoo/gn-f405/default.px4board
+++ b/boards/flywoo/gn-f405/default.px4board
@@ -7,6 +7,7 @@ CONFIG_BOARD_SERIAL_URT6="/dev/ttyS2"
 CONFIG_BOARD_SERIAL_TEL2="/dev/ttyS1"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_BMP280=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_MPU6000=y
diff --git a/boards/hkust/nxt-dual/default.px4board b/boards/hkust/nxt-dual/default.px4board
index 8844879555..cab44e7812 100644
--- a/boards/hkust/nxt-dual/default.px4board
+++ b/boards/hkust/nxt-dual/default.px4board
@@ -12,6 +12,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/hkust/nxt-v1/default.px4board b/boards/hkust/nxt-v1/default.px4board
index bc04f1ce94..6261079e0d 100644
--- a/boards/hkust/nxt-v1/default.px4board
+++ b/boards/hkust/nxt-v1/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/holybro/durandal-v1/default.px4board b/boards/holybro/durandal-v1/default.px4board
index 362eaff742..ecc1bc8101 100644
--- a/boards/holybro/durandal-v1/default.px4board
+++ b/boards/holybro/durandal-v1/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/holybro/kakutef7/default.px4board b/boards/holybro/kakutef7/default.px4board
index bcaedd9cf2..8aaa037cc5 100644
--- a/boards/holybro/kakutef7/default.px4board
+++ b/boards/holybro/kakutef7/default.px4board
@@ -9,6 +9,7 @@ CONFIG_BOARD_SERIAL_TEL2="/dev/ttyS1"
 CONFIG_BOARD_SERIAL_RC="/dev/ttyS4"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_BMP280=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_ICM20689=y
diff --git a/boards/holybro/kakuteh7/default.px4board b/boards/holybro/kakuteh7/default.px4board
index 270caa7b65..dbf8aeb7f2 100644
--- a/boards/holybro/kakuteh7/default.px4board
+++ b/boards/holybro/kakuteh7/default.px4board
@@ -9,6 +9,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/holybro/kakuteh7mini/default.px4board b/boards/holybro/kakuteh7mini/default.px4board
index b49f1cfa01..eb465e0027 100644
--- a/boards/holybro/kakuteh7mini/default.px4board
+++ b/boards/holybro/kakuteh7mini/default.px4board
@@ -9,6 +9,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/holybro/kakuteh7v2/default.px4board b/boards/holybro/kakuteh7v2/default.px4board
index d9d432beec..45fd6e11eb 100644
--- a/boards/holybro/kakuteh7v2/default.px4board
+++ b/boards/holybro/kakuteh7v2/default.px4board
@@ -9,6 +9,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/holybro/pix32v5/default.px4board b/boards/holybro/pix32v5/default.px4board
index 45bd43ee4f..0093ee2dcf 100644
--- a/boards/holybro/pix32v5/default.px4board
+++ b/boards/holybro/pix32v5/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/matek/h743-mini/default.px4board b/boards/matek/h743-mini/default.px4board
index 64e472d770..22efafcfdb 100644
--- a/boards/matek/h743-mini/default.px4board
+++ b/boards/matek/h743-mini/default.px4board
@@ -7,6 +7,7 @@ CONFIG_BOARD_SERIAL_TEL3="/dev/ttyS5"
 CONFIG_BOARD_SERIAL_TEL4="/dev/ttyS6"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_DPS310=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_ICM20602=y
diff --git a/boards/matek/h743-slim/default.px4board b/boards/matek/h743-slim/default.px4board
index ce5f222229..71024fea5c 100644
--- a/boards/matek/h743-slim/default.px4board
+++ b/boards/matek/h743-slim/default.px4board
@@ -7,6 +7,7 @@ CONFIG_BOARD_SERIAL_TEL3="/dev/ttyS5"
 CONFIG_BOARD_SERIAL_TEL4="/dev/ttyS6"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_DPS310=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
diff --git a/boards/matek/h743/default.px4board b/boards/matek/h743/default.px4board
index ce5f222229..dcd9fad5ae 100644
--- a/boards/matek/h743/default.px4board
+++ b/boards/matek/h743/default.px4board
@@ -8,6 +8,7 @@ CONFIG_BOARD_SERIAL_TEL4="/dev/ttyS6"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_DPS310=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_ICM20602=y
diff --git a/boards/modalai/fc-v1/default.px4board b/boards/modalai/fc-v1/default.px4board
index 2a33d1ddb8..e9104cfaee 100644
--- a/boards/modalai/fc-v1/default.px4board
+++ b/boards/modalai/fc-v1/default.px4board
@@ -12,6 +12,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/modalai/fc-v2/default.px4board b/boards/modalai/fc-v2/default.px4board
index 6fa7ebabaf..15b94d5dc8 100644
--- a/boards/modalai/fc-v2/default.px4board
+++ b/boards/modalai/fc-v2/default.px4board
@@ -12,6 +12,7 @@ CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_ICM42688P=y
diff --git a/boards/mro/ctrl-zero-classic/default.px4board b/boards/mro/ctrl-zero-classic/default.px4board
index d07a7c1fc9..665fbc8626 100644
--- a/boards/mro/ctrl-zero-classic/default.px4board
+++ b/boards/mro/ctrl-zero-classic/default.px4board
@@ -12,6 +12,7 @@ CONFIG_DRIVERS_BAROMETER_DPS310=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/mro/ctrl-zero-f7-oem/default.px4board b/boards/mro/ctrl-zero-f7-oem/default.px4board
index b0701855bf..9fe340b87f 100644
--- a/boards/mro/ctrl-zero-f7-oem/default.px4board
+++ b/boards/mro/ctrl-zero-f7-oem/default.px4board
@@ -10,6 +10,7 @@ CONFIG_DRIVERS_BAROMETER_DPS310=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/mro/ctrl-zero-f7/default.px4board b/boards/mro/ctrl-zero-f7/default.px4board
index 074a4f2ca0..d09ca8a5b5 100644
--- a/boards/mro/ctrl-zero-f7/default.px4board
+++ b/boards/mro/ctrl-zero-f7/default.px4board
@@ -9,6 +9,7 @@ CONFIG_DRIVERS_BAROMETER_DPS310=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/mro/ctrl-zero-h7-oem/default.px4board b/boards/mro/ctrl-zero-h7-oem/default.px4board
index 549ba99b73..ec6c59888d 100644
--- a/boards/mro/ctrl-zero-h7-oem/default.px4board
+++ b/boards/mro/ctrl-zero-h7-oem/default.px4board
@@ -10,6 +10,7 @@ CONFIG_DRIVERS_BAROMETER_DPS310=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/mro/ctrl-zero-h7/default.px4board b/boards/mro/ctrl-zero-h7/default.px4board
index c32ae10fc0..7bb96d3aca 100644
--- a/boards/mro/ctrl-zero-h7/default.px4board
+++ b/boards/mro/ctrl-zero-h7/default.px4board
@@ -10,6 +10,7 @@ CONFIG_DRIVERS_BAROMETER_DPS310=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/mro/pixracerpro/default.px4board b/boards/mro/pixracerpro/default.px4board
index d4b1d29370..40301136e2 100644
--- a/boards/mro/pixracerpro/default.px4board
+++ b/boards/mro/pixracerpro/default.px4board
@@ -9,6 +9,7 @@ CONFIG_DRIVERS_BAROMETER_DPS310=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/mro/x21-777/default.px4board b/boards/mro/x21-777/default.px4board
index 5a51a776dd..361aca4484 100644
--- a/boards/mro/x21-777/default.px4board
+++ b/boards/mro/x21-777/default.px4board
@@ -9,6 +9,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/mro/x21/default.px4board b/boards/mro/x21/default.px4board
index 7c20626001..8640bff623 100644
--- a/boards/mro/x21/default.px4board
+++ b/boards/mro/x21/default.px4board
@@ -10,6 +10,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/nxp/fmuk66-e/default.px4board b/boards/nxp/fmuk66-e/default.px4board
index b8bc5cc7f4..749e62e5d5 100644
--- a/boards/nxp/fmuk66-e/default.px4board
+++ b/boards/nxp/fmuk66-e/default.px4board
@@ -11,6 +11,7 @@ CONFIG_DRIVERS_BAROMETER_MPL3115A2=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DISTANCE_SENSOR_SRF05=y
diff --git a/boards/nxp/fmuk66-v3/default.px4board b/boards/nxp/fmuk66-v3/default.px4board
index e53ae83c88..268e89c70f 100644
--- a/boards/nxp/fmuk66-v3/default.px4board
+++ b/boards/nxp/fmuk66-v3/default.px4board
@@ -11,6 +11,7 @@ CONFIG_DRIVERS_BAROMETER_MPL3115A2=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DISTANCE_SENSOR_SRF05=y
diff --git a/boards/omnibus/f4sd/default.px4board b/boards/omnibus/f4sd/default.px4board
index 287d991659..8d3c58a7f0 100644
--- a/boards/omnibus/f4sd/default.px4board
+++ b/boards/omnibus/f4sd/default.px4board
@@ -7,6 +7,7 @@ CONFIG_BOARD_SERIAL_URT6="/dev/ttyS2"
 CONFIG_BOARD_SERIAL_TEL2="/dev/ttyS1"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_BMP280=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_ICM20602=y
diff --git a/boards/px4/fmu-v2/default.px4board b/boards/px4/fmu-v2/default.px4board
index 59d1c02a73..f0753941e0 100644
--- a/boards/px4/fmu-v2/default.px4board
+++ b/boards/px4/fmu-v2/default.px4board
@@ -9,6 +9,7 @@ CONFIG_BOARD_SERIAL_TEL2="/dev/ttyS2"
 CONFIG_BOARD_SERIAL_TEL4="/dev/ttyS6"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_MS5611=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_MPU6000=y
 CONFIG_DRIVERS_IMU_ST_L3GD20=y
diff --git a/boards/px4/fmu-v3/default.px4board b/boards/px4/fmu-v3/default.px4board
index 1ced5b7873..5c24d0e731 100644
--- a/boards/px4/fmu-v3/default.px4board
+++ b/boards/px4/fmu-v3/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/px4/fmu-v4/default.px4board b/boards/px4/fmu-v4/default.px4board
index 0e1bb8f547..ab106bdb7b 100644
--- a/boards/px4/fmu-v4/default.px4board
+++ b/boards/px4/fmu-v4/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/px4/fmu-v4pro/default.px4board b/boards/px4/fmu-v4pro/default.px4board
index f722e649ef..ff8455c285 100644
--- a/boards/px4/fmu-v4pro/default.px4board
+++ b/boards/px4/fmu-v4pro/default.px4board
@@ -12,6 +12,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/px4/fmu-v5/default.px4board b/boards/px4/fmu-v5/default.px4board
index bfb8b6af46..beb8658321 100644
--- a/boards/px4/fmu-v5/default.px4board
+++ b/boards/px4/fmu-v5/default.px4board
@@ -10,6 +10,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/px4/fmu-v5/protected.px4board b/boards/px4/fmu-v5/protected.px4board
index e5371b2a7d..fd3b7bb522 100644
--- a/boards/px4/fmu-v5/protected.px4board
+++ b/boards/px4/fmu-v5/protected.px4board
@@ -6,6 +6,7 @@ CONFIG_COMMON_OPTICAL_FLOW=n
 CONFIG_COMMON_TELEMETRY=n
 CONFIG_DRIVERS_CAMERA_CAPTURE=n
 CONFIG_DRIVERS_CAMERA_TRIGGER=n
+CONFIG_DRIVERS_CDCACM_AUTOSTART=n
 CONFIG_DRIVERS_IRLOCK=n
 CONFIG_DRIVERS_PCA9685_PWM_OUT=n
 CONFIG_DRIVERS_PWM_INPUT=n
diff --git a/boards/px4/fmu-v5x/default.px4board b/boards/px4/fmu-v5x/default.px4board
index 2bede96763..46d04d8cc2 100644
--- a/boards/px4/fmu-v5x/default.px4board
+++ b/boards/px4/fmu-v5x/default.px4board
@@ -14,6 +14,7 @@ CONFIG_DRIVERS_BAROMETER_MS5611=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/px4/fmu-v6c/default.px4board b/boards/px4/fmu-v6c/default.px4board
index 148e6326fb..51484f7903 100644
--- a/boards/px4/fmu-v6c/default.px4board
+++ b/boards/px4/fmu-v6c/default.px4board
@@ -10,6 +10,7 @@ CONFIG_DRIVERS_BAROMETER_MS5611=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/px4/fmu-v6u/default.px4board b/boards/px4/fmu-v6u/default.px4board
index 50b4f6ba3d..e733dccc35 100644
--- a/boards/px4/fmu-v6u/default.px4board
+++ b/boards/px4/fmu-v6u/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/px4/fmu-v6x/default.px4board b/boards/px4/fmu-v6x/default.px4board
index f197a254a0..7b9488933f 100644
--- a/boards/px4/fmu-v6x/default.px4board
+++ b/boards/px4/fmu-v6x/default.px4board
@@ -14,6 +14,7 @@ CONFIG_DRIVERS_BAROMETER_INVENSENSE_ICP201XX=y
 CONFIG_DRIVERS_BAROMETER_MS5611=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/px4/fmu-v6xrt/default.px4board b/boards/px4/fmu-v6xrt/default.px4board
index 029462aa48..17434ed18d 100644
--- a/boards/px4/fmu-v6xrt/default.px4board
+++ b/boards/px4/fmu-v6xrt/default.px4board
@@ -13,6 +13,7 @@ CONFIG_DRIVERS_BAROMETER_INVENSENSE_ICP201XX=y
 CONFIG_DRIVERS_BAROMETER_MS5611=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/raspberrypi/pico/default.px4board b/boards/raspberrypi/pico/default.px4board
index a5727267a1..410f8d4788 100644
--- a/boards/raspberrypi/pico/default.px4board
+++ b/boards/raspberrypi/pico/default.px4board
@@ -6,6 +6,7 @@ CONFIG_BOARD_SERIAL_GPS1="/dev/ttyS2"
 CONFIG_BOARD_SERIAL_TEL1="/dev/ttyS1"
 CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_DRIVERS_BAROMETER_BMP280=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_IMU_INVENSENSE_MPU9250=y
 CONFIG_DRIVERS_MAGNETOMETER_HMC5883=y
diff --git a/boards/siyi/n7/default.px4board b/boards/siyi/n7/default.px4board
index cb42f3530c..3364cf55f7 100644
--- a/boards/siyi/n7/default.px4board
+++ b/boards/siyi/n7/default.px4board
@@ -10,6 +10,7 @@ CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
 CONFIG_DRIVERS_HEATER=y
diff --git a/boards/sky-drones/smartap-airlink/default.px4board b/boards/sky-drones/smartap-airlink/default.px4board
index 4bb3c7cc40..7613792e2c 100644
--- a/boards/sky-drones/smartap-airlink/default.px4board
+++ b/boards/sky-drones/smartap-airlink/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/spracing/h7extreme/default.px4board b/boards/spracing/h7extreme/default.px4board
index c6c28729a7..aa97905609 100644
--- a/boards/spracing/h7extreme/default.px4board
+++ b/boards/spracing/h7extreme/default.px4board
@@ -4,6 +4,7 @@ CONFIG_DRIVERS_ADC_BOARD_ADC=y
 CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
diff --git a/boards/thepeach/k1/default.px4board b/boards/thepeach/k1/default.px4board
index 39fefb9b68..d760279546 100644
--- a/boards/thepeach/k1/default.px4board
+++ b/boards/thepeach/k1/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/thepeach/r1/default.px4board b/boards/thepeach/r1/default.px4board
index 39fefb9b68..d760279546 100644
--- a/boards/thepeach/r1/default.px4board
+++ b/boards/thepeach/r1/default.px4board
@@ -11,6 +11,7 @@ CONFIG_COMMON_BAROMETERS=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DIFFERENTIAL_PRESSURE=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
diff --git a/boards/uvify/core/default.px4board b/boards/uvify/core/default.px4board
index bdc591a281..8c144aa606 100644
--- a/boards/uvify/core/default.px4board
+++ b/boards/uvify/core/default.px4board
@@ -10,6 +10,7 @@ CONFIG_DRIVERS_BAROMETER_MS5611=y
 CONFIG_DRIVERS_BATT_SMBUS=y
 CONFIG_DRIVERS_CAMERA_CAPTURE=y
 CONFIG_DRIVERS_CAMERA_TRIGGER=y
+CONFIG_DRIVERS_CDCACM_AUTOSTART=y
 CONFIG_COMMON_DISTANCE_SENSOR=y
 CONFIG_DRIVERS_DSHOT=y
 CONFIG_DRIVERS_GPS=y
diff --git a/platforms/nuttx/src/px4/common/CMakeLists.txt b/platforms/nuttx/src/px4/common/CMakeLists.txt
index ab16e77a0e..c74810b03a 100644
--- a/platforms/nuttx/src/px4/common/CMakeLists.txt
+++ b/platforms/nuttx/src/px4/common/CMakeLists.txt
@@ -38,7 +38,6 @@ if(NOT PX4_BOARD MATCHES "io-v2")
 		board_crashdump.c
 		board_dma_alloc.c
 		board_fat_dma_alloc.c
-		cdc_acm_check.cpp
 		console_buffer.cpp
 		cpuload.cpp
 		gpio.c
diff --git a/platforms/nuttx/src/px4/common/px4_init.cpp b/platforms/nuttx/src/px4/common/px4_init.cpp
index ceeb79bd72..aa11b1cf70 100644
--- a/platforms/nuttx/src/px4/common/px4_init.cpp
+++ b/platforms/nuttx/src/px4/common/px4_init.cpp
@@ -62,8 +62,6 @@
 #include <px4_platform/board_ctrl.h>
 #endif
 
-extern void cdcacm_init(void);
-
 #if !defined(CONFIG_BUILD_FLAT)
 typedef CODE void (*initializer_t)(void);
 extern initializer_t _sinit;
@@ -196,10 +194,6 @@ int px4_platform_init()
 
 	px4_log_initialize();
 
-#if defined(CONFIG_SYSTEM_CDCACM) && defined(CONFIG_BUILD_FLAT)
-	cdcacm_init();
-#endif
-
 	return PX4_OK;
 }
 
diff --git a/platforms/nuttx/src/px4/common/px4_layer.cmake b/platforms/nuttx/src/px4/common/px4_layer.cmake
index 9199a95289..de8fe5a0c6 100644
--- a/platforms/nuttx/src/px4/common/px4_layer.cmake
+++ b/platforms/nuttx/src/px4/common/px4_layer.cmake
@@ -2,7 +2,6 @@
 
 add_library(px4_layer
 		${KERNEL_SRCS}
-		cdc_acm_check.cpp
 		${PX4_SOURCE_DIR}/platforms/common/Serial.cpp
 		SerialImpl.cpp
 	)
diff --git a/platforms/nuttx/src/px4/common/px4_protected_layers.cmake b/platforms/nuttx/src/px4/common/px4_protected_layers.cmake
index c55fe568d3..5fbbef1644 100644
--- a/platforms/nuttx/src/px4/common/px4_protected_layers.cmake
+++ b/platforms/nuttx/src/px4/common/px4_protected_layers.cmake
@@ -6,7 +6,6 @@ add_library(px4_layer
 	board_fat_dma_alloc.c
 	tasks.cpp
 	console_buffer_usr.cpp
-	cdc_acm_check.cpp
 	${PX4_SOURCE_DIR}/platforms/posix/src/px4/common/print_load.cpp
 	${PX4_SOURCE_DIR}/platforms/posix/src/px4/common/cpuload.cpp
 	px4_userspace_init.cpp
diff --git a/platforms/nuttx/src/px4/common/px4_userspace_init.cpp b/platforms/nuttx/src/px4/common/px4_userspace_init.cpp
index f5d9882ca2..82cace9e4d 100644
--- a/platforms/nuttx/src/px4/common/px4_userspace_init.cpp
+++ b/platforms/nuttx/src/px4/common/px4_userspace_init.cpp
@@ -44,8 +44,6 @@
 #include <uORB/uORB.h>
 #include <sys/boardctl.h>
 
-extern void cdcacm_init(void);
-
 extern "C" void px4_userspace_init(void)
 {
 	hrt_init();
@@ -55,8 +53,4 @@ extern "C" void px4_userspace_init(void)
 	px4::WorkQueueManagerStart();
 
 	uorb_start();
-
-#if defined(CONFIG_SYSTEM_CDCACM)
-	cdcacm_init();
-#endif
 }
diff --git a/src/drivers/cdcacm_autostart/CMakeLists.txt b/src/drivers/cdcacm_autostart/CMakeLists.txt
new file mode 100644
index 0000000000..8e78ba1cd2
--- /dev/null
+++ b/src/drivers/cdcacm_autostart/CMakeLists.txt
@@ -0,0 +1,40 @@
+############################################################################
+#
+#   Copyright (c) 2023 PX4 Development Team. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in
+#    the documentation and/or other materials provided with the
+#    distribution.
+# 3. Neither the name PX4 nor the names of its contributors may be
+#    used to endorse or promote products derived from this software
+#    without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+# AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+############################################################################
+px4_add_module(
+	MODULE drivers__cdcacm_autostart
+	MAIN cdcacm_autostart
+	COMPILE_FLAGS
+		# -DDEBUG_BUILD
+	SRCS
+		cdcacm_autostart.cpp
+	)
diff --git a/src/drivers/cdcacm_autostart/Kconfig b/src/drivers/cdcacm_autostart/Kconfig
new file mode 100644
index 0000000000..aabfaad483
--- /dev/null
+++ b/src/drivers/cdcacm_autostart/Kconfig
@@ -0,0 +1,6 @@
+menuconfig DRIVERS_CDCACM_AUTOSTART
+	bool "cdcacm_autostart"
+	default n
+	depends on MODULES_MAVLINK
+	---help---
+		Enable support for cdcacm_autostart
\ No newline at end of file
diff --git a/src/drivers/cdcacm_autostart/cdcacm_autostart.cpp b/src/drivers/cdcacm_autostart/cdcacm_autostart.cpp
new file mode 100644
index 0000000000..d7ebbddd6b
--- /dev/null
+++ b/src/drivers/cdcacm_autostart/cdcacm_autostart.cpp
@@ -0,0 +1,663 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2023 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#if defined(CONFIG_SYSTEM_CDCACM)
+
+#include "cdcacm_autostart.h"
+
+__BEGIN_DECLS
+#include <arch/board/board.h>
+#include <builtin/builtin.h>
+
+extern int sercon_main(int c, char **argv);
+extern int serdis_main(int c, char **argv);
+__END_DECLS
+
+#include <px4_platform_common/shutdown.h>
+
+#define USB_DEVICE_PATH "/dev/ttyACM0"
+
+#if defined(CONFIG_SERIAL_PASSTHRU_UBLOX)
+#  undef SERIAL_PASSTHRU_UBLOX_DEV
+#  if defined(CONFIG_SERIAL_PASSTHRU_GPS1) && defined(CONFIG_BOARD_SERIAL_GPS1)
+#    define SERIAL_PASSTHRU_UBLOX_DEV CONFIG_BOARD_SERIAL_GPS1
+#  elif defined(CONFIG_SERIAL_PASSTHRU_GPS2)&& defined(CONFIG_BOARD_SERIAL_GPS2)
+#    define SERIAL_PASSTHRU_UBLOX_DEV CONFIG_BOARD_SERIAL_GPS2
+#  elif defined(CONFIG_SERIAL_PASSTHRU_GPS3)&& defined(CONFIG_BOARD_SERIAL_GPS3)
+#    define SERIAL_PASSTHRU_UBLOX_DEV CONFIG_BOARD_SERIAL_GPS3
+#  elif defined(CONFIG_SERIAL_PASSTHRU_GPS4)&& defined(CONFIG_BOARD_SERIAL_GPS4)
+#    define SERIAL_PASSTHRU_UBLOX_DEV CONFIG_BOARD_SERIAL_GPS4
+#  elif defined(CONFIG_SERIAL_PASSTHRU_GPS5) && defined(CONFIG_BOARD_SERIAL_GPS5)
+#    define SERIAL_PASSTHRU_UBLOX_DEV CONFIG_BOARD_SERIAL_GPS5
+#  endif
+#  if !defined(SERIAL_PASSTHRU_UBLOX_DEV)
+#    error "CONFIG_SERIAL_PASSTHRU_GPSn and CONFIG_BOARD_SERIAL_GPSn must be defined"
+#  endif
+#endif
+
+CdcAcmAutostart::CdcAcmAutostart() :
+	ModuleParams(nullptr),
+	ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::lp_default)
+{}
+
+CdcAcmAutostart::~CdcAcmAutostart()
+{
+	PX4_INFO("Stopping CDC/ACM autostart");
+
+	if (_ttyacm_fd >= 0) {
+		px4_close(_ttyacm_fd);
+	}
+
+	ScheduleClear();
+}
+
+int CdcAcmAutostart::Start()
+{
+	PX4_INFO("Starting CDC/ACM autostart");
+	UpdateParams(true);
+
+	ScheduleNow();
+
+	return PX4_OK;
+}
+
+void CdcAcmAutostart::Run()
+{
+	if (should_exit()) {
+		exit_and_cleanup();
+		return;
+	}
+
+	UpdateParams();
+
+	run_state_machine();
+}
+
+void CdcAcmAutostart::run_state_machine()
+{
+	_reschedule_time = 500_ms;
+	_vbus_present = (board_read_VBUS_state() == PX4_OK);
+
+	// If the hardware supports RESET lockout that has nArmed ANDed with VBUS
+	// vbus_sense may drop during a param save which uses
+	// BOARD_INDICATE_EXTERNAL_LOCKOUT_STATE to prevent external resets
+	// while writing the params.  If we are not armed and nARMRED is low
+	// we are in such a lock out so ignore changes on VBUS_SENSE during this
+	// time.
+#if defined(BOARD_GET_EXTERNAL_LOCKOUT_STATE)
+
+	if (BOARD_GET_EXTERNAL_LOCKOUT_STATE() == 0) {
+		_vbus_present = _vbus_present_prev;
+		ScheduleDelayed(500_ms);
+		return;
+	}
+
+#endif
+
+	// Do not reconfigure USB while flying
+	actuator_armed_s report;
+	_actuator_armed_sub.copy(&report);
+
+	if (report.armed) {
+		_vbus_present_prev = _vbus_present;
+
+	} else {
+
+		switch (_state) {
+		case UsbAutoStartState::disconnected:
+			state_disconnected();
+			break;
+
+		case UsbAutoStartState::connecting:
+			state_connecting();
+			break;
+
+		case UsbAutoStartState::connected:
+			state_connected();
+			break;
+
+		case UsbAutoStartState::disconnecting:
+			state_disconnecting();
+			break;
+		}
+	}
+
+	_vbus_present_prev = _vbus_present;
+	ScheduleDelayed(_reschedule_time);
+}
+
+void CdcAcmAutostart::state_connected()
+{
+	if (!_vbus_present && !_vbus_present_prev && (_active_protocol == UsbProtocol::mavlink)) {
+		PX4_DEBUG("lost vbus!");
+		sched_lock();
+		static const char app[] {"mavlink"};
+		static const char *stop_argv[] {"mavlink", "stop", "-d", USB_DEVICE_PATH, NULL};
+		exec_builtin(app, (char **)stop_argv, NULL, 0);
+		sched_unlock();
+		_state = UsbAutoStartState::disconnecting;
+	}
+}
+
+void CdcAcmAutostart::state_disconnected()
+{
+	if (_vbus_present && _vbus_present_prev) {
+		PX4_DEBUG("starting sercon");
+
+		if (sercon_main(0, nullptr) == EXIT_SUCCESS) {
+			_state = UsbAutoStartState::connecting;
+			PX4_DEBUG("state connecting");
+			_reschedule_time = 1_s;
+		}
+
+	} else if (_vbus_present && !_vbus_present_prev) {
+		// USB just connected, try again soon
+		_reschedule_time = 100_ms;
+	}
+}
+
+void CdcAcmAutostart::state_connecting()
+{
+	int bytes_available = 0;
+#if defined(CONFIG_SERIAL_PASSTHRU_UBLOX)
+	struct termios uart_config;
+	speed_t baudrate;
+#endif
+
+	if (!_vbus_present) {
+		PX4_DEBUG("No VBUS");
+		goto fail;
+	}
+
+	if (_ttyacm_fd < 0) {
+		PX4_DEBUG("opening port");
+		_ttyacm_fd = px4_open(USB_DEVICE_PATH, O_RDONLY | O_NONBLOCK);
+	}
+
+	if (_ttyacm_fd < 0) {
+		PX4_DEBUG("can't open port");
+		goto fail;
+	}
+
+	if (_sys_usb_auto.get() == 2) {
+		PX4_INFO("Starting mavlink on %s (SYS_USB_AUTO=2)", USB_DEVICE_PATH);
+
+		if (start_mavlink()) {
+			_state = UsbAutoStartState::connected;
+			_active_protocol = UsbProtocol::mavlink;
+
+		} else {
+			_state = UsbAutoStartState::disconnecting;
+			_reschedule_time = 100_ms;
+		}
+
+		return;
+
+	} else if (_sys_usb_auto.get() == 0) {
+		// Do nothing
+		_state = UsbAutoStartState::connected;
+		_active_protocol = UsbProtocol::none;
+		return;
+	}
+
+	// Otherwise autodetect
+
+	if ((px4_ioctl(_ttyacm_fd, FIONREAD, &bytes_available) != PX4_OK) ||
+	    (bytes_available < 3)) {
+		PX4_DEBUG("bytes_available: %d", bytes_available);
+		// Return back to connecting state to check again soon
+		return;
+	}
+
+	// Non-blocking read
+	_bytes_read = px4_read(_ttyacm_fd, _buffer, sizeof(_buffer));
+
+#if defined(DEBUG_BUILD)
+
+	if (_bytes_read > 0) {
+		fprintf(stderr, "%d bytes\n", _bytes_read);
+
+		for (int i = 0; i < _bytes_read; i++) {
+			fprintf(stderr, "|%X", _buffer[i]);
+		}
+
+		fprintf(stderr, "\n");
+	}
+
+#endif // DEBUG_BUILD
+
+	if (_bytes_read <= 0) {
+		PX4_DEBUG("no _bytes_read");
+		// Return back to connecting state to check again soon
+		return;
+	}
+
+#if defined(CONFIG_SERIAL_PASSTHRU_UBLOX)
+	// Get the baudrate for serial passthru before closing the port.
+	tcgetattr(_ttyacm_fd, &uart_config);
+	baudrate = cfgetspeed(&uart_config);
+#endif
+	PX4_DEBUG("_bytes_read %d", _bytes_read);
+	px4_close(_ttyacm_fd);
+	_ttyacm_fd = -1;
+
+	// Parse for mavlink reboot command
+	if (scan_buffer_for_mavlink_reboot()) {
+		// Reboot incoming. Return without rescheduling.
+		return;
+	}
+
+	// Parse for mavlink heartbeats (v1 and v2).
+	if (scan_buffer_for_mavlink_heartbeat()) {
+		if (start_mavlink()) {
+			_state = UsbAutoStartState::connected;
+			_active_protocol = UsbProtocol::mavlink;
+
+		} else {
+			_state = UsbAutoStartState::disconnecting;
+			_reschedule_time = 100_ms;
+		}
+
+		return;
+	}
+
+	// Parse for carriage returns indicating someone is trying to use the nsh.
+	if (scan_buffer_for_carriage_returns()) {
+		if (start_nsh()) {
+			_state = UsbAutoStartState::connected;
+			_active_protocol = UsbProtocol::nsh;
+
+		} else {
+			_state = UsbAutoStartState::disconnecting;
+			_reschedule_time = 100_ms;
+		}
+
+		return;
+	}
+
+#if defined(CONFIG_SERIAL_PASSTHRU_UBLOX)
+
+	// Parse for ublox start of packet byte sequence.
+	if (scan_buffer_for_ublox_bytes()) {
+		if (start_ublox_serial_passthru(baudrate)) {
+			_state = UsbAutoStartState::connected;
+			_active_protocol = UsbProtocol::ublox;
+
+		} else {
+			_state = UsbAutoStartState::disconnecting;
+			_reschedule_time = 100_ms;
+		}
+
+		return;
+	}
+
+#endif
+
+	return;
+
+fail:
+	PX4_DEBUG("fail...");
+
+	// VBUS not present, open failed
+	if (_ttyacm_fd >= 0) {
+		px4_close(_ttyacm_fd);
+		_ttyacm_fd = -1;
+	}
+
+	_state = UsbAutoStartState::disconnecting;
+}
+
+void CdcAcmAutostart::state_disconnecting()
+{
+	PX4_DEBUG("state_disconnecting");
+
+	if (_ttyacm_fd > 0) {
+		px4_close(_ttyacm_fd);
+	}
+
+	// Disconnect serial
+	serdis_main(0, NULL);
+	_state = UsbAutoStartState::disconnected;
+	_active_protocol = UsbProtocol::none;
+}
+
+bool CdcAcmAutostart::scan_buffer_for_mavlink_reboot()
+{
+	bool rebooting = false;
+
+	// Mavlink reboot/shutdown command
+	// COMMAND_LONG (#76) with command MAV_CMD_PREFLIGHT_REBOOT_SHUTDOWN (246)
+	static constexpr int MAVLINK_COMMAND_LONG_MIN_LENGTH = 41;
+
+	if (_bytes_read < MAVLINK_COMMAND_LONG_MIN_LENGTH) {
+		return rebooting;
+	}
+
+	// scan buffer for mavlink COMMAND_LONG
+	for (int i = 0; i < _bytes_read - MAVLINK_COMMAND_LONG_MIN_LENGTH; i++) {
+		if ((_buffer[i] == 0xFE)        // Mavlink v1 start byte
+		    && (_buffer[i + 5] == 76)   //  76=0x4C COMMAND_LONG
+		    && (_buffer[i + 34] == 246) // 246=0xF6 MAV_CMD_PREFLIGHT_REBOOT_SHUTDOWN
+		   ) {
+			// mavlink v1 COMMAND_LONG
+			//  buffer[0]: start byte (0xFE for mavlink v1)
+			//  buffer[3]: SYSID
+			//  buffer[4]: COMPID
+			//  buffer[5]: message id (COMMAND_LONG 76=0x4C)
+			//  buffer[6-10]: COMMAND_LONG param 1 (little endian float)
+			//  buffer[34]: COMMAND_LONG command MAV_CMD_PREFLIGHT_REBOOT_SHUTDOWN (246/0xF6)
+			float param1_raw = 0;
+			memcpy(&param1_raw, &_buffer[i + 6], 4);
+			int param1 = roundf(param1_raw);
+
+			PX4_INFO("%s: Mavlink MAV_CMD_PREFLIGHT_REBOOT_SHUTDOWN param 1: %d (SYSID:%d COMPID:%d)",
+				 USB_DEVICE_PATH, param1, _buffer[i + 3], _buffer[i + 4]);
+
+			if (param1 == 1) {
+				// 1: Reboot autopilot
+				rebooting = true;
+				px4_reboot_request(REBOOT_REQUEST, 0);
+
+			} else if (param1 == 2) {
+				// 2: Shutdown autopilot
+#if defined(BOARD_HAS_POWER_CONTROL)
+				rebooting = true;
+				px4_shutdown_request(0);
+#endif // BOARD_HAS_POWER_CONTROL
+
+			} else if (param1 == 3) {
+				// 3: Reboot autopilot and keep it in the bootloader until upgraded.
+				rebooting = true;
+				px4_reboot_request(REBOOT_TO_BOOTLOADER, 0);
+			}
+		}
+	}
+
+	return rebooting;
+}
+
+bool CdcAcmAutostart::scan_buffer_for_mavlink_heartbeat()
+{
+	static constexpr int MAVLINK_HEARTBEAT_MIN_LENGTH = 9;
+	bool start_mavlink = false;
+
+	if (_bytes_read < MAVLINK_HEARTBEAT_MIN_LENGTH) {
+		return start_mavlink;
+	}
+
+	// scan buffer for mavlink HEARTBEAT (v1 & v2)
+	for (int i = 0; i < _bytes_read - MAVLINK_HEARTBEAT_MIN_LENGTH; i++) {
+		if ((_buffer[i] == 0xFE) && (_buffer[i + 1] == 9) && (_buffer[i + 5] == 0)) {
+			// mavlink v1 HEARTBEAT
+			//  buffer[0]: start byte (0xFE for mavlink v1)
+			//  buffer[1]: length (9 for HEARTBEAT)
+			//  buffer[3]: SYSID
+			//  buffer[4]: COMPID
+			//  buffer[5]: mavlink message id (0 for HEARTBEAT)
+			PX4_INFO("%s: launching mavlink (HEARTBEAT v1 from SYSID:%d COMPID:%d)",
+				 USB_DEVICE_PATH, _buffer[i + 3], _buffer[i + 4]);
+			start_mavlink = true;
+
+		} else if ((_buffer[i] == 0xFD) && (_buffer[i + 1] == 9)
+			   && (_buffer[i + 7] == 0) && (_buffer[i + 8] == 0) && (_buffer[i + 9] == 0)) {
+			// mavlink v2 HEARTBEAT
+			//  buffer[0]: start byte (0xFD for mavlink v2)
+			//  buffer[1]: length (9 for HEARTBEAT)
+			//  buffer[5]: SYSID
+			//  buffer[6]: COMPID
+			//  buffer[7:9]: mavlink message id (0 for HEARTBEAT)
+			PX4_INFO("%s: launching mavlink (HEARTBEAT v2 from SYSID:%d COMPID:%d)",
+				 USB_DEVICE_PATH, _buffer[i + 5], _buffer[i + 6]);
+			start_mavlink = true;
+		}
+	}
+
+	return start_mavlink;
+}
+
+bool CdcAcmAutostart::scan_buffer_for_carriage_returns()
+{
+	bool start_nsh = false;
+
+	if (_bytes_read < 3) {
+		return start_nsh;
+	}
+
+	// nshterm (3 carriage returns)
+	// scan buffer looking for 3 consecutive carriage returns (0xD)
+	for (int i = 1; i < _bytes_read - 1; i++) {
+		if (_buffer[i - 1] == 0xD && _buffer[i] == 0xD && _buffer[i + 1] == 0xD) {
+			PX4_INFO("%s: launching nshterm", USB_DEVICE_PATH);
+			start_nsh = true;
+			break;
+		}
+	}
+
+	return start_nsh;
+}
+
+bool CdcAcmAutostart::scan_buffer_for_ublox_bytes()
+{
+	bool success = false;
+
+	if (_bytes_read < 4) {
+		return success;
+	}
+
+	// scan buffer looking for 0xb5 0x62 which indicates the start of a packet
+	for (int i = 0; i < _bytes_read; i++) {
+		bool ub = _buffer[i] == 0xb5 && _buffer[i + 1] == 0x62;
+
+		if (ub && ((_buffer[i + 2 ] == 0x6 && (_buffer[i + 3 ] == 0xb8 || _buffer[i + 3 ] == 0x13)) ||
+			   (_buffer[i + 2 ] == 0xa && _buffer[i + 3 ] == 0x4))) {
+			PX4_INFO("%s: launching ublox serial passthru", USB_DEVICE_PATH);
+			success = true;
+			break;
+		}
+	}
+
+	return success;
+}
+
+bool CdcAcmAutostart::start_mavlink()
+{
+	bool success = false;
+	char mavlink_mode_string[3];
+	snprintf(mavlink_mode_string, sizeof(mavlink_mode_string), "%ld", _usb_mav_mode.get());
+	static const char *argv[] {"mavlink", "start", "-d", USB_DEVICE_PATH, "-m", mavlink_mode_string, nullptr};
+
+	if (execute_process((char **)argv) > 0) {
+		success = true;
+	}
+
+	return success;
+}
+
+bool CdcAcmAutostart::start_nsh()
+{
+	bool success = false;
+	static const char *argv[] {"nshterm", USB_DEVICE_PATH, nullptr};
+
+	if (execute_process((char **)argv) > 0) {
+		success = true;
+	}
+
+	return success;
+}
+
+#if defined(CONFIG_SERIAL_PASSTHRU_UBLOX)
+bool CdcAcmAutostart::start_ublox_serial_passthru(speed_t baudrate)
+{
+	bool success = false;
+	char baudstring[16];
+	snprintf(baudstring, sizeof(baudstring), "%ld", baudrate);
+
+	// Stop the GPS driver first
+	static const char *gps_argv[] {"gps", "stop", nullptr};
+	static const char *passthru_argv[] {"serial_passthru", "start", "-t", "-b", baudstring, "-e", USB_DEVICE_PATH, "-d", SERIAL_PASSTHRU_UBLOX_DEV, nullptr};
+
+	if (execute_process((char **)gps_argv) > 0) {
+		if (execute_process((char **)passthru_argv) > 0) {
+			success = true;
+		}
+	}
+
+	return success;
+}
+#endif
+
+int CdcAcmAutostart::execute_process(char **argv)
+{
+	int pid = -1;
+	sched_lock();
+
+	pid = exec_builtin(argv[0], argv, nullptr, 0);
+
+	sched_unlock();
+	return pid;
+}
+
+int CdcAcmAutostart::task_spawn(int argc, char *argv[])
+{
+	CdcAcmAutostart *instance = new CdcAcmAutostart();
+
+	if (!instance) {
+		PX4_ERR("alloc failed");
+		return -1;
+	}
+
+	int ret = instance->Start();
+
+	if (ret != PX4_OK) {
+		delete instance;
+		return ret;
+	}
+
+	_object.store(instance);
+	_task_id = task_id_is_work_queue;
+
+	return ret;
+}
+
+void CdcAcmAutostart::UpdateParams(const bool force)
+{
+	if (_parameter_update_sub.updated() || force) {
+		parameter_update_s param_update;
+		_parameter_update_sub.copy(&param_update);
+		ModuleParams::updateParams();
+	}
+}
+
+int CdcAcmAutostart::custom_command(int argc, char *argv[])
+{
+	return print_usage("unknown command");
+}
+
+int CdcAcmAutostart::print_status()
+{
+	const char *state = "";
+	const char *protocol = "";
+
+	switch (_state) {
+	case UsbAutoStartState::disconnected:
+		state = "disconnected";
+		break;
+
+	case UsbAutoStartState::connecting:
+		state = "connecting";
+		break;
+
+	case UsbAutoStartState::connected:
+		state = "connected";
+		break;
+
+	case UsbAutoStartState::disconnecting:
+		state = "disconnecting";
+		break;
+	}
+
+	switch (_active_protocol) {
+	case UsbProtocol::none:
+		protocol = "none";
+		break;
+
+	case UsbProtocol::mavlink:
+		protocol = "mavlink";
+		break;
+
+	case UsbProtocol::nsh:
+		protocol = "nsh";
+		break;
+
+	case UsbProtocol::ublox:
+		protocol = "ublox";
+		break;
+	}
+
+	PX4_INFO("Running");
+	PX4_INFO("State: %s", state);
+	PX4_INFO("Protocol: %s", protocol);
+	return PX4_OK;
+}
+
+int CdcAcmAutostart::print_usage(const char *reason)
+{
+	if (reason) {
+		printf("%s\n\n", reason);
+	}
+
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
+### Description
+This module listens on USB and auto-configures the protocol depending on the bytes received.
+The supported protocols are: MAVLink, nsh, and ublox serial passthrough. If the parameter SYS_USB_AUTO=2
+the module will only try to start mavlink as long as the USB VBUS is detected. Otherwise it will spin
+and continue to check for VBUS and start mavlink once it is detected.
+)DESCR_STR");
+
+	PRINT_MODULE_USAGE_NAME("cdcacm_autostart", "system");
+	PRINT_MODULE_USAGE_COMMAND("start");
+	PRINT_MODULE_USAGE_DEFAULT_COMMANDS();
+
+	return 0;
+}
+
+#endif
+
+extern "C" __EXPORT int cdcacm_autostart_main(int argc, char *argv[])
+{
+#if defined(CONFIG_SYSTEM_CDCACM)
+	return CdcAcmAutostart::main(argc, argv);
+#endif
+	return 1;
+}
diff --git a/src/drivers/cdcacm_autostart/cdcacm_autostart.h b/src/drivers/cdcacm_autostart/cdcacm_autostart.h
new file mode 100644
index 0000000000..a0e1271515
--- /dev/null
+++ b/src/drivers/cdcacm_autostart/cdcacm_autostart.h
@@ -0,0 +1,124 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2023 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <px4_platform_common/module.h>
+#include <px4_platform_common/module_params.h>
+#include <px4_platform_common/px4_work_queue/ScheduledWorkItem.hpp>
+#include <uORB/SubscriptionInterval.hpp>
+#include <uORB/topics/parameter_update.h>
+#include <uORB/topics/actuator_armed.h>
+
+#include <termios.h>
+
+using namespace time_literals;
+
+class CdcAcmAutostart : public ModuleBase<CdcAcmAutostart>, public ModuleParams, public px4::ScheduledWorkItem
+{
+public:
+	CdcAcmAutostart();
+	~CdcAcmAutostart() override;
+
+	/** @see ModuleBase */
+	static int task_spawn(int argc, char *argv[]);
+
+	/** @see ModuleBase */
+	static int custom_command(int argc, char *argv[]);
+
+	/** @see ModuleBase */
+	static int print_usage(const char *reason = nullptr);
+
+	/** @see ModuleBase */
+	int print_status() override;
+
+	int Start();
+
+private:
+
+	enum class UsbAutoStartState {
+		disconnected,
+		connecting,
+		connected,
+		disconnecting,
+	};
+
+	enum class UsbProtocol {
+		none,
+		mavlink,
+		nsh,
+		ublox,
+	};
+
+	void Run() override;
+
+	void UpdateParams(const bool force = false);
+
+	void run_state_machine();
+
+	void state_disconnected();
+	void state_connecting();
+	void state_connected();
+	void state_disconnecting();
+
+	bool scan_buffer_for_mavlink_reboot();
+	bool scan_buffer_for_mavlink_heartbeat();
+	bool scan_buffer_for_carriage_returns();
+	bool scan_buffer_for_ublox_bytes();
+
+	bool start_mavlink();
+	bool start_nsh();
+#if defined(CONFIG_SERIAL_PASSTHRU_UBLOX)
+	bool start_ublox_serial_passthru(speed_t baudrate);
+#endif
+	int execute_process(char **argv);
+
+	uORB::Subscription	_actuator_armed_sub{ORB_ID(actuator_armed)};
+	uORB::SubscriptionInterval _parameter_update_sub{ORB_ID(parameter_update), 500_ms};
+
+	UsbAutoStartState _state{UsbAutoStartState::disconnected};
+	UsbProtocol _active_protocol{UsbProtocol::none};
+	bool _vbus_present = false;
+	bool _vbus_present_prev = false;
+	int _ttyacm_fd = -1;
+
+	char _buffer[80] = {};
+	int _bytes_read = 0;
+
+	uint32_t _reschedule_time = 0;
+
+	DEFINE_PARAMETERS(
+		(ParamInt<px4::params::SYS_USB_AUTO>) _sys_usb_auto,
+		(ParamInt<px4::params::USB_MAV_MODE>) _usb_mav_mode
+	)
+};
diff --git a/src/drivers/cdcacm_autostart/cdcacm_autostart_params.c b/src/drivers/cdcacm_autostart/cdcacm_autostart_params.c
new file mode 100644
index 0000000000..3dcc16ad2f
--- /dev/null
+++ b/src/drivers/cdcacm_autostart/cdcacm_autostart_params.c
@@ -0,0 +1,68 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2024 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * Enable USB autostart
+ *
+ * @value 0 Disabled
+ * @value 1 Auto-detect
+ * @value 2 MAVLink
+ *
+ * @reboot_required true
+ *
+ * @group CDCACM
+ */
+PARAM_DEFINE_INT32(SYS_USB_AUTO, 2);
+
+/**
+ * Specify USB MAVLink mode
+ *
+ * @value 0 normal
+ * @value 1 custom
+ * @value 2 onboard
+ * @value 3 osd
+ * @value 4 magic
+ * @value 5 config
+ * @value 6 iridium
+ * @value 7 minimal
+ * @value 8 extvision
+ * @value 9 extvisionmin
+ * @value 10 gimbal
+ * @value 11 onboard_low_bandwidth
+ * @value 12 uavionix
+ *
+ * @reboot_required true
+ *
+ * @group CDCACM
+ */
+PARAM_DEFINE_INT32(USB_MAV_MODE, 2);
\ No newline at end of file
diff --git a/src/modules/mavlink/mavlink_main.cpp b/src/modules/mavlink/mavlink_main.cpp
index 722c1254f6..186bfe7936 100644
--- a/src/modules/mavlink/mavlink_main.cpp
+++ b/src/modules/mavlink/mavlink_main.cpp
@@ -2094,11 +2094,6 @@ Mavlink::task_main(int argc, char *argv[])
 
 		/* USB has no baudrate, but use a magic number for 'fast' */
 		_baudrate = 2000000;
-
-		if (_mode == MAVLINK_MODE_COUNT) {
-			_mode = MAVLINK_MODE_CONFIG;
-		}
-
 		_ftp_on = true;
 		_is_usb_uart = true;
 
@@ -2223,11 +2218,24 @@ Mavlink::task_main(int argc, char *argv[])
 
 	/* open the UART device after setting the instance, as it might block */
 	if (get_protocol() == Protocol::SERIAL) {
-		_uart_fd = mavlink_open_uart(_baudrate, _device_name, _flow_control);
 
-		if (_uart_fd < 0) {
-			PX4_ERR("could not open %s", _device_name);
-			return PX4_ERROR;
+		// NOTE: we attempt to open the port multiple times due to sercon returning before
+		// the port is ready to be opened. This avoids needing to sleep() after sercon_main.
+		int attempts = 0;
+		static const int max_attempts = 3;
+
+		while (_uart_fd < 0) {
+			_uart_fd = mavlink_open_uart(_baudrate, _device_name, _flow_control);
+			attempts++;
+
+			if (_uart_fd < 0 && attempts < max_attempts) {
+				PX4_ERR("could not open %s, retrying", _device_name);
+				px4_usleep(1_s);
+
+			} else if (_uart_fd < 0) {
+				PX4_ERR("failed to open %s after %d attempts, exiting!", _device_name, attempts);
+				return PX4_ERROR;
+			}
 		}
 	}
 
diff --git a/src/systemcmds/serial_passthru/Kconfig b/src/systemcmds/serial_passthru/Kconfig
index 4c4cb107d8..aabd8c59fe 100644
--- a/src/systemcmds/serial_passthru/Kconfig
+++ b/src/systemcmds/serial_passthru/Kconfig
@@ -11,7 +11,7 @@ config SERIAL_PASSTHRU_UBLOX
 	bool "Detect and Auto Connect on U-Center messages"
 	default n
 	---help---
-		This option will enable the cdc_acm_check to launch
+		This option will enable the cdcacm_autostart to launch
 		The passthru driver.
 
 #
-- 
2.48.1

